README.md
# jsnpp
Java SNPP API

This library can be used by your java applictions to send messages via Simple Network Paging Protocol (SNPP) servers and handle response from those servers.
## API Javadoc
* http://dtseiler.github.io/jsnpp/docs/api/


ChangeLog
ChangeLog for jsnpp - https://github.com/dtseiler/jsnpp

*jsnpp-1.0.0 (16 Feb 2016)
	New Features:
	* Added support for Connection and Socket timeouts.
	* Welcome @mkellstrand to the project.

*jsnpp-0.3.1 (01 Feb 2006)
	Bug Fixes:
	* Fixed errors in LOGI with password due to misuse of String.concat().
	  Thanks cainey at users.sourceforge.net.

	New Features:
	* Accepting 230 success code from LOGI for naughty Hylafax SNPP servers.
	  Thanks cainey at users.sourceforge.net.
	* Now building zip files for distribution.

*jsnpp-0.3.0 (14 Oct 2005)
	Bug Fixes:
	* Changed build process to put version in jsnpp jar file name, like all
	  good libs should.  This one will be jsnpp-0.3.0.jar.
	
	New Features:
	* Added in password support for LOGI command, just call
	  Message.setLogin(login, password).

*jsnpp-0.2.0 (11 Aug 2005)
	Bug Fixes:
	* Fixed distfile creation so it is tarred inside a release dir.

	New Features:
	* Added support for SNPP Level 2 commands.
	* Added support for SNPP Level 3 SEND respones 860 & 890.
	* Added Message.getSENDResponse() function for user to get the server
	  response to the SEND command.
	* API documentation available at http://jsnpp.sourceforge.net/api/

*jsnpp-0.1.0 (20 May 2005)
	* Initial creation.


tests/jsnpp/AllTests.java
/**
 *	JSNPP - Java SNPP API.
 *	Copyright (C) 2005  Don Seiler <don@seiler.us>
 *	
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation; either
 *	version 2.1 of the License, or (at your option) any later version.
 *	
 *	This library is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	Lesser General Public License for more details.
 *	
 *	You should have received a copy of the GNU Lesser General Public
 *	License along with this library; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package jsnpp;

import junit.framework.Test;
import junit.framework.TestSuite;

/**
 * Runs all tests
 *
 * @author Don Seiler <don@NOSPAM.seiler.us>
 * @version $Revision$
 */
public class AllTests {
	public static Test suite() {
		TestSuite suite = new TestSuite();

		suite.addTestSuite(Level1Test.class);
		//suite.addTestSuite(Level2Test.class);
		//suite.addTestSuite(Level3Test.class);

		return suite;
	}
}


tests/jsnpp/Level1Test.java
/**
 *	JSNPP - Java SNPP API.
 *	Copyright (C) 2005  Don Seiler <don@seiler.us>
 *	
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation; either
 *	version 2.1 of the License, or (at your option) any later version.
 *	
 *	This library is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	Lesser General Public License for more details.
 *	
 *	You should have received a copy of the GNU Lesser General Public
 *	License along with this library; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package jsnpp;

import java.io.IOException;
import java.net.UnknownHostException;
import junit.framework.TestCase;

/**
 * The SNPP Level 1 test case.
 *
 * @author Don Seiler <don@NOSPAM.seiler.us>
 * @version $Revision$
 */
public class Level1Test extends TestCase {
	Message m = null;
	
	public Level1Test (String n) {
		super(n);
	}

	protected void setUp() {
		m = new Message("test.mysnppserver.com", 444, "don", "The system is DOWN.");
	}

	protected void tearDown() {
	}

	public void testSendPage() {
		try {
			m.send();
			System.out.println("Response was: " + m.getSENDResponse());
			System.out.println("Level 1 page sent.");
		} catch (UnknownHostException e) {
			System.err.println("Unknown host: " + e.getMessage());
		} catch (IOException e) {
			System.err.println("IO error: " + e.getMessage());
		} catch (Exception e) {
			System.err.println("Error sending message: " + e.getMessage());
		}
	}
}


tests/jsnpp/Level2Test.java
/**
 *	JSNPP - Java SNPP API.
 *	Copyright (C) 2005  Don Seiler <don@seiler.us>
 *	
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation; either
 *	version 2.1 of the License, or (at your option) any later version.
 *	
 *	This library is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	Lesser General Public License for more details.
 *	
 *	You should have received a copy of the GNU Lesser General Public
 *	License along with this library; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package jsnpp;

import java.io.IOException;
import java.net.UnknownHostException;
import junit.framework.TestCase;

/**
 * The SNPP Level 2 test case.
 *
 * @author Don Seiler <don@NOSPAM.seiler.us>
 * @version $Revision$
 */
public class Level2Test extends TestCase {
	Message m = null;
	String[] data = { "This", "is being", "sent on separate", "lines." };
	
	public Level2Test (String n) {
		super(n);
	}

	protected void setUp() {
		m = new Message("test.mysnppserver.com", 444, "don");
		//m.setLevel(2);
		m.setSubject("The Subject");
		//m.setCoverageArea(2);
		m.setAlertOverride(true);
		m.setData(data);
	}

	protected void tearDown() {
	}

	public void testSendPage() {
		try {
			m.send();
			System.out.println("Level 2 page sent.");
		} catch (UnknownHostException e) {
			System.err.println("Unknown host: " + e.getMessage());
		} catch (IOException e) {
			System.err.println("IO error: " + e.getMessage());
		} catch (Exception e) {
			System.err.println("Error sending message: " + e.getMessage());
		}
	}
}



src/jsnpp/Message.java
/**
 *	JSNPP - Java SNPP API.
 *	Copyright (C) 2005  Don Seiler <don@seiler.us>
 *	
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation; either
 *	version 2.1 of the License, or (at your option) any later version.
 *	
 *	This library is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	Lesser General Public License for more details.
 *	
 *	You should have received a copy of the GNU Lesser General Public
 *	License along with this library; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package jsnpp;

import java.io.IOException;
import java.net.UnknownHostException;

/**
 * SNPP Message Class
 *
 * @author Don Seiler <don@seiler.us>
 * @version $Id$
 */
public class Message {

	// Maximum SNPP Level Supported
	public static final int MAX_SNPP_LEVEL = 3;

	// Message states
	public static final int STATE_UNDEF = -1;
	public static final int STATE_CONN = 0;
	public static final int STATE_PAGE = 11;
	public static final int STATE_MESS = 12;
	public static final int STATE_SEND = 13;
	public static final int STATE_QUIT = 14;
	public static final int STATE_DATA = 21;
	public static final int STATE_LOGI = 22;
	public static final int STATE_LEVE = 23;
	public static final int STATE_COVE = 24;
	public static final int STATE_SUBJ = 25;
	public static final int STATE_HOLD = 26;
	public static final int STATE_CALL = 27;
	public static final int STATE_ALER = 28;


	/** Connection to server */
	private Connection conn = null;


	/** Message state */
	private int state = STATE_UNDEF;


	/** Pager (recipient) */
	private String pager = null;


	/** Message */
	private String message = null;


	/** 
	 * Data 
	 *
	 * A message cannot send both MESS and DATA.
	 */
	private String[] data = null;


	/** CallerIdentifier */
	private String callerIdentifier = null;


	/** Subject */
	private String subject = null;


	/** Login */
	private String login = null;


	/** Password */
	private String password = null;


	/** Service Level */
	private int serviceLevel = -1;


	/** Coverage Area */
	private int coverageArea = -1;


	/** Hold Until YYMMDDHHMISS (+/- GMT diff) */
	private String holdUntil = null;


	/** Alert Override 0-DoNotAlert or 1-Alert */
	private boolean alertOverride = false;


	/** Level */
	private int level = MAX_SNPP_LEVEL;


	/** Track this response */
	private String sendResponse = null;

	/** Socket connection timeout in millis.  Zero is an infinite timeout. **/
	int socketConnectTimeout = 0;

	/** Socket read timeout in millis. Zero is an infinite timeout. **/
	private int socketInputTimeout = 0;
	

	/** Constructor */
	public Message() {
	}


	/** Constructor */
	public Message (String host,
					int port,
					String pager) {
		setPager(pager);
		
		setConnectionInfo(host, port);
	}


	/** Constructor */
	public Message (String host,
					int port,
					String pager,
					String message) {
		setPager(pager);
		setMessage(message);
		
		setConnectionInfo(host, port);
	}


	/** 
	 * Constructor 
	 *
	 * @deprecated
	 */
	public Message (String host,
					int port,
					String sender,
					String pager,
					String message) {
		setCallerIdentifier(sender);
		setPager(pager);
		setMessage(message);

		setConnectionInfo(host, port);
	}


	/** Returns state */
	public int getState () {
		return state;
	}


	/** Returns callerIdentifier */
	public String getCallerIdentifier () {
		return callerIdentifier;
	}


	/**
	 * Returns recipient 
	 *
	 * @deprecated
	 */
	public String getRecipient () {
		return getPager();
	}


	/** Returns pager */
	public String getPager() {
		return pager;
	}


	/** Returns message */
	public String getMessage () {
		return message;
	}


	/** Sets connection info */
	public void setConnectionInfo(String host, int port) {
		conn = new Connection(host, port);
		state = STATE_CONN;
	}

	/** Get the socket SO_TIMEOUT (read) timeout, in milliseconds **/
	public int getSocketInputTimeout() {
		return socketInputTimeout;
	}

	/**
	 * Enable/disable the socket SO_TIMEOUT with the specified (read) timeout, in milliseconds.
	 * The timeout must be > 0. A timeout of zero is interpreted as an infinite timeout.
	 *
	 * @param socketInputTimeout
     	*/
	public void setSocketInputTimeout(int socketInputTimeout) {
		if (socketInputTimeout >= 0) {
			this.socketInputTimeout = socketInputTimeout;
		}
	}

	/** Get the server connection timeout, in milliseconds **/
	public int getSocketConnectTimeout() {
		return socketConnectTimeout;
	}

	/**
	 * Set the server connection timeout value in milliseconds.
	 * A timeout of zero is interpreted as an infinite timeout.
	 *
	 * @param socketConnectTimeout
     	*/
	public void setSocketConnectTimeout(int socketConnectTimeout) {
		if (socketConnectTimeout >= 0) {
			this.socketConnectTimeout = socketConnectTimeout;
		}
	}

	/** Sets state */
	public void setState(int state) {
		this.state = state;
	}


	/** 
	 * Sets sender 
	 *
	 * @deprecated
	 */
	public void setSender(String sender) {
		setCallerIdentifier(sender);
	}


	/** Sets callerIdentifer */
	public void setCallerIdentifier(String callerIdentifier) {
		this.callerIdentifier = callerIdentifier;
	}


	/** 
	 * Sets recipient 
	 *
	 * @deprecated
	 */
	public void setRecipient(String recipient) {
		setPager(recipient);
	}


	/** Sets pager */
	public void setPager(String pager) {
		this.pager = pager;
	}


	/** Sets message */
	public void setMessage(String message) {
		// Check for \n chars
		this.message = message;
	}


	/** Returns SNPP level of this message */
	public int getLevel() {
		return level;
	}


	/** Sets SNPP level of this message */
	public void setLevel(int level) {
		if (level <= MAX_SNPP_LEVEL)
			this.level = level;
		else
			this.level = MAX_SNPP_LEVEL;
	}


	/** Sets subject */
	public void setSubject(String subject) {
		this.subject = subject;
	}


	/** Returns subject */
	public String getSubject() {
		return subject;
	}


	/** Sets data */
	public void setData(String[] data) {
		this.data = data;
	}


	/** Returns data */
	public String[] getData() {
		return data;
	}


	/** Sets alert override */
	public void setAlertOverride(boolean alertOverride) {
		this.alertOverride = alertOverride;
	}


	/** Returns alert override */
	public boolean getAlertOverride() {
		return alertOverride;
	}


	/**
	 * Sets holdUntil
	 *
	 * Must be in YYMMDDHHMISS format
	 */
	public void setHoldUntil(String holdUntil) {
		this.holdUntil = holdUntil;
	}


	/** Returns "hold until" value */
	public String getHoldUntil() {
		return holdUntil;
	}


	/** Sets login */
	public void setLogin(String login) {
		this.login = login;
	}


	/** Sets login with password */
	public void SetLogin(String login, String password) {
		this.login = login;
		this.password = password;
	}


	/** Returns login */
	public String getLogin() {
		return login;
	}


	/** Sets coverage area */
	public void setCoverageArea(int coverageArea) {
		this.coverageArea = coverageArea;
	}


	/** Returns coverage area */
	public int getCoverageArea() {
		return coverageArea;
	}


	/** Sets service level */
	public void setServiceLevel(int serviceLevel) {
		this.serviceLevel = serviceLevel;
	}


	/** Returns service level */
	public int getServiceLevel() {
		return serviceLevel;
	}


	/** Returns response from SEND command */
	public String getSENDResponse() {
		return sendResponse;
	}


	//
	// Function to format and send messages
	//

	/**
	 * Connects to server and sends message.
	 *
	 * @throws	SocketCloseException	if error while closing socket
	 * @throws	IOException		If bad socket I/O with SNPP server.
	 * @throws	Exception		If bad or unknown server response received.
	 */
	public void send() throws SocketCloseException, IOException, Exception {
		handleResponse(conn.connect(socketConnectTimeout, socketInputTimeout));
	}


	/**
	 * Handles response from server.
	 *
	 * We ignore all 500 messages (Command Not Implemented) from optional Level 2
	 * functions (optional as defined in RFC 1861).
	 */
	private void handleResponse(String response) throws IOException, Exception {

		// XXX PAGE should be sent AFTER LEVE, ALER, HOLD, COVE
		
		switch (state) {
			case STATE_CONN:
				if (response.startsWith("220"))
					if (level >= 2)
						sendLOGICommand();
					else
						sendPAGECommand();
				else
					errorOut(response);
				break;
				
				
			case STATE_LOGI:
                // 230 is a success response that Hylafax server software uses, for some reason.
				if (response.startsWith("230") || response.startsWith("250") || response.startsWith("500"))
					sendCOVECommand();
				else
					errorOut(response);
				break;
				
				
			case STATE_COVE:
				if (response.startsWith("250") || response.startsWith("500"))
					sendLEVECommand();
				else
					errorOut(response);
				break;
				
				
			case STATE_LEVE:
				if (response.startsWith("250") || response.startsWith("500"))
					sendHOLDCommand();
				else
					errorOut(response);
				break;
				
				
			case STATE_HOLD:
				if (response.startsWith("250") || response.startsWith("500"))
					sendALERCommand();
				else
					errorOut(response);
				break;
				
				
			case STATE_ALER:
				if (response.startsWith("250") || response.startsWith("500")) 
					sendPAGECommand();
				else
					errorOut(response);
				break;
				

			case STATE_PAGE:
				if (response.startsWith("250")) {
					if (level >= 2)
						sendCALLCommand();
					else
						sendMESSCommand();
				} else
					errorOut(response);
				break;

				
			case STATE_CALL:
				if (response.startsWith("250") || response.startsWith("500"))
					sendSUBJCommand();
				else
					errorOut(response);
				break;
				
				
			case STATE_SUBJ:
				if (response.startsWith("250") || response.startsWith("500"))
					sendMESSCommand();
				else
					errorOut(response);
				break;

				
			case STATE_MESS:
				if (response.startsWith("250"))
					sendSENDCommand();
				else
					errorOut(response);
				break;
				
			case STATE_DATA:
				if (response.startsWith("354"))
					sendDATAData();
				else if (response.startsWith("250"))
					sendSENDCommand();
				else
					errorOut(response);
				break;

				
			case STATE_SEND:
				// Accept level 1-3 success messages
				sendResponse = response;
				if (response.startsWith("250")
						|| response.startsWith("860") 
						|| response.startsWith("960"))
					sendQUITCommand();
				else
					errorOut(response);
				break;

				
			case STATE_QUIT:
				if (response.startsWith("221"))
					conn.close();
				else
					errorOut(response);
				break;
		}
	}


	private void sendPAGECommand() throws IOException, Exception {
		state = STATE_PAGE;
		String msg = "PAGE " + pager;
		handleResponse(conn.send(msg));
	}


	private void sendCALLCommand() throws IOException, Exception {
		if (callerIdentifier != null) {
			state = STATE_CALL;
			String msg = "CALL " + callerIdentifier;
			handleResponse(conn.send(msg));
		} else {
			sendMESSCommand();
		}
	}


	private void sendCOVECommand() throws IOException, Exception {
		if (coverageArea > 0) {
			state = STATE_COVE;
			String msg = "COVE " + coverageArea;
			handleResponse(conn.send(msg));
		} else {
			sendLEVECommand();
		}
	}


	private void sendLEVECommand() throws IOException, Exception {
		if (serviceLevel > 0) {
			state = STATE_LEVE;
			String msg = "LEVE " + serviceLevel;
			handleResponse(conn.send(msg));
		} else {
			sendHOLDCommand();
		}
	}


	private void sendHOLDCommand() throws IOException, Exception {
		if (holdUntil != null) {
			state = STATE_HOLD;
			String msg = "HOLD " + holdUntil;
			handleResponse(conn.send(msg));
		} else {
			sendALERCommand();
		}
	}


	private void sendALERCommand() throws IOException, Exception {
		if (alertOverride) {
			state = STATE_ALER;
			handleResponse(conn.send("ALER"));
		} else {
			sendPAGECommand();
		}
	}


	private void sendSUBJCommand() throws IOException, Exception {
		if (subject != null) {
			state = STATE_SUBJ;
			String msg = "SUBJ " + subject;
			handleResponse(conn.send(msg));
		} else {
			sendMESSCommand();
		}
	}


	private void sendMESSCommand() throws IOException, Exception {
		state = STATE_MESS;
		if ((level >= 2)&&(message == null)) {
			sendDATACommand();
		} else {
			String msg = "MESS " + message;
			handleResponse(conn.send(msg));
		}
	}


	private void sendSENDCommand() throws IOException, Exception {
		state = STATE_SEND;
		handleResponse(conn.send("SEND"));
	}


	private void sendQUITCommand() throws IOException, Exception {
		state = STATE_QUIT;
		handleResponse(conn.send("QUIT"));
	}


	private void sendLOGICommand() throws IOException, Exception {
		if (login != null) {
			state = STATE_LOGI;
            String msg = "LOGI "+ login + ((password != null) ? " " + password : "");
			handleResponse(conn.send(msg));
		} else {
			sendCOVECommand();
		}
	}


	private void sendDATACommand() throws IOException, Exception {
		state = STATE_DATA;
		handleResponse(conn.send("DATA"));
	}


	private void sendDATAData() throws IOException, Exception {
		for (int i=0; i < data.length; i++) {
			conn.send(data[i], false);
		}
		handleResponse(conn.send("."));
	}


	private void errorOut(String badResponse) throws Exception {
		throw new Exception (badResponse);
	}
}


src/jsnpp/Connection.java
/**
 *	JSNPP - Java SNPP API.
 *	Copyright (C) 2005  Don Seiler <don@seiler.us>
 *	
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation; either
 *	version 2.1 of the License, or (at your option) any later version.
 *	
 *	This library is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	Lesser General Public License for more details.
 *	
 *	You should have received a copy of the GNU Lesser General Public
 *	License along with this library; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package jsnpp;

import java.io.PrintWriter;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.net.Socket;
import java.net.InetSocketAddress;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;

/**
 * SNPP Connection Class
 *
 * @author Don Seiler <don@seiler.us>
 * @version $Id$
 */
public class Connection {

	private Socket socket = null;
	private PrintWriter out = null;
	private BufferedReader in = null;

	private String host = null;
	private int port = 444;

	/** Constructor */
	public Connection(String host, int port) {
		this.host = host;

		if (port > 0)
			this.port = port;
	}


	/**
	 * Connects to host/port of SNPP server.
	 *
	 * This method will create the socket to the SNPP server, and return the
	 * response code.  A successful connection will return a String beginning
	 * with "220", similar to this:
	 *
	 * 220 QuickPage v3.3 SNPP server ready at Tue May 17 11:48:12 2005
	 *
	 * @return Response of SNPP server to connection.
	 */
	public String connect() throws UnknownHostException, IOException {
		return connect(0, 0);
	}


	/**
	 * Connects to host/port of SNPP server, with specified timeout settings.
	 *
	 * This method will create the socket to the SNPP server, and return the
	 * response code.  A successful connection will return a String beginning
	 * with "220", similar to this:
	 *
	 * 220 QuickPage v3.3 SNPP server ready at Tue May 17 11:48:12 2005
	 *
	 * @param socketConnectTimeout	Timeout in milliseconds for connection attempt. A timeout of zero is interpreted as an infinite timeout.
	 * @param socketInputTimeout	Timeout in milliseconds to wait on responses from the SNPP server. A timeout of zero is interpreted as an infinite timeout.
	 * @return Response of SNPP server to connection.
	 */
	public String connect(int socketConnectTimeout, int socketInputTimeout)
			throws UnknownHostException, IOException, SocketTimeoutException {
		
		Socket socket = new Socket();
		socket.connect(new InetSocketAddress(host, port), socketConnectTimeout);
		socket.setSoTimeout(socketInputTimeout);

		out = new PrintWriter(socket.getOutputStream(), true);
		in = new BufferedReader(new InputStreamReader(socket.getInputStream()));

		// This will contain the immediate response
		return in.readLine();
	}

	/**
	 * Closes connection to SNPP server.
	 * If any Exceptions are thrown while closing the streams or socket,
	 * then the first Exception encountered will be re-thrown.
	 *
	 * @throws SocketCloseException		Exception thrown while closing the socket
     	*/
	public void close() throws SocketCloseException {

		SocketCloseException sce = null;

		if (out != null) {
			try {
				out.close();
			} catch (Exception e) {
				if (sce == null) {
					sce = new SocketCloseException(e);
				}
			}
		}

		if (in != null) {
			try {
				in.close();
			} catch (Exception e) {
				if (sce == null) {
					sce = new SocketCloseException(e);
				}
			}
		}

		if (socket != null) {
			try {
				socket.close();
			} catch (Exception e) {
				if (sce == null) {
					sce = new SocketCloseException(e);
				}
			}
		}

		if (sce != null) {
			throw sce;
		}
	}

	/** Sends data to SNPP server */
	public String send(String data) throws IOException {
		return send(data, true);
	}
	
	public String send(String data, boolean wait) throws IOException {
		String response = null;
		
		// Send command to server
		out.println(data);
		//System.out.println(data);
		
		// Read response
		if (wait) {
			response = in.readLine();
			//System.out.println(response);
		}
		
		// Return response, or null
		return response;
	}
}


src/jsnpp/SocketCloseException.java
/**
 *	JSNPP - Java SNPP API.
 *	Copyright (C) 2005  Don Seiler <don@seiler.us>
 *
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation; either
 *	version 2.1 of the License, or (at your option) any later version.
 *
 *	This library is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	Lesser General Public License for more details.
 *
 *	You should have received a copy of the GNU Lesser General Public
 *	License along with this library; if not, write to the Free Software
 *	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package jsnpp;

import java.io.IOException;

/**
 * SocketCloseException thrown when closing streams and sockets
 *
 * @author Mike Kellstrand
 * @version $Id$
 */
public class SocketCloseException extends IOException {

    public SocketCloseException(Exception e) {
        super(e);
    }
}


examples/jsnpp/SNPPExample.java
/*
 * JSNPP - Java SNPP API.
 * Copyright (C) 2005  Don Seiler <don@seiler.us>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

package jsnpp;

import jsnpp.Message;
import java.io.IOException;
import java.net.UnknownHostException;

public class SNPPExample {

	public static void main(String[] args) {
		Message m = new Message("test.mysnppserver.com", 444, "don", "The system is DOWN.");
		try {
			m.send();
		} catch (UnknownHostException e) {
			System.err.println("Unknown host: " + e.getMessage());
		} catch (IOException e) {
			System.err.println("IO error: " + e.getMessage());
		} catch (Exception e) {
			System.err.println("Error sending message: " + e.getMessage());
		}
	}
}
